src/vm/eventpipe.cpp

```diff
@@ -343,34 +336,33 @@ void EventPipe::Disable(EventPipeSessionID id)
 
     SetupThread();
 
-    // Only perform the disable operation if the session ID
-    // matches the current active session.
-    if (id != (EventPipeSessionID)s_pSession)
+    if (id == 0 || !Enabled())
         return;
 
     // Don't block GC during clean-up.
     GCX_PREEMP();
 
-    EventPipe::RunWithCallbackPostponed(
-        [&](EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
-        {
-            DisableInternal(reinterpret_cast<EventPipeSessionID>(s_pSession), pEventPipeProviderCallbackDataQueue);
-        }
-    );
+    RunWithCallbackPostponed([&](EventPipeProviderCallbackDataQueue *pEventPipeProviderCallbackDataQueue) {
+        EventPipeSession *pSession = reinterpret_cast<EventPipeSession *>(id);
+        if (!s_pSessions->Contains(pSession))
+            return;
+        DisableInternal(*pSession, pEventPipeProviderCallbackDataQueue);
+        // TODO: Remove session from map, delete session, delete file.
+    });
 }
 
-void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
+void EventPipe::DisableInternal(EventPipeSession &session, EventPipeProviderCallbackDataQueue* pEventPipeProviderCallbackDataQueue)
 {
     CONTRACTL
     {
         THROWS;
         GC_TRIGGERS;
         MODE_ANY;
-        PRECONDITION(GetLock()->OwnedByCurrentThread());
+        PRECONDITION(IsLockOwnedByCurrentThread());
     }
     CONTRACTL_END;
 
-    if (s_pConfig != NULL && s_pConfig->Enabled())
+    if (s_pConfig != nullptr)
     {
         // Disable the profiler.
         SampleProfiler::Disable();
```

Comment:
We should not disable the SampleProfiler when only one session disable?

Reply:
Yes, this is incomplete. I'll push a commit with @noahfalk suggestion.